src/mono/mono/component/diagnostics_server.c

```diff
@@ -19,6 +28,291 @@ static MonoComponentDiagnosticsServer fn_table = {
 	&ds_server_disable
 };
 
+#else /* !defined (HOST_WASM) || defined (DISABLE_THREADS) */
+
+static bool
+ds_server_wasm_init (void);
+
+static bool
+ds_server_wasm_shutdown (void);
+
+static void
+ds_server_wasm_pause_for_diagnostics_monitor (void);
+
+static void
+ds_server_wasm_disable (void);
+
+static MonoComponentDiagnosticsServer fn_table = {
+	{ MONO_COMPONENT_ITF_VERSION, &diagnostics_server_available },
+	&ds_server_wasm_init,
+	&ds_server_wasm_shutdown,
+	&ds_server_wasm_pause_for_diagnostics_monitor,
+	&ds_server_wasm_disable,
+};
+
+typedef struct _MonoWasmDiagnosticServerOptions {
+	int32_t suspend; /* set from JS! */
+	MonoCoopSem suspend_resume;
+} MonoWasmDiagnosticServerOptions;
+
+static MonoWasmDiagnosticServerOptions wasm_ds_options;
+static pthread_t ds_thread_id;
+
+extern void
+mono_wasm_diagnostic_server_on_runtime_server_init (MonoWasmDiagnosticServerOptions *out_options);
+
+EMSCRIPTEN_KEEPALIVE void
+mono_wasm_diagnostic_server_resume_runtime_startup (void);
+
+static bool
+ds_server_wasm_init (void)
+{
+	/* called on the main thread when the runtime is sufficiently initialized */
+	mono_coop_sem_init (&wasm_ds_options.suspend_resume, 0);
+	mono_wasm_diagnostic_server_on_runtime_server_init(&wasm_ds_options);
+	return true;
+}
+
+
+static bool
+ds_server_wasm_shutdown (void)
+{
+	return true;
+}
+
+static void
+ds_server_wasm_pause_for_diagnostics_monitor (void)
+{
+	/* wait until the DS receives a resume */
+	if (wasm_ds_options.suspend) {
+		const guint timeout = 50;
+		const guint warn_threshold = 5000;
+		guint cumulative_timeout = 0;
+		while (true) {
+			MonoSemTimedwaitRet res = mono_coop_sem_timedwait (&wasm_ds_options.suspend_resume, timeout, MONO_SEM_FLAGS_ALERTABLE);
+			if (res == MONO_SEM_TIMEDWAIT_RET_SUCCESS || res == MONO_SEM_TIMEDWAIT_RET_ALERTED)
+				break;
+			else {
+				/* timed out */
+				cumulative_timeout += timeout;
```

Comment:
I used to print a warning to the JS console here, but it's not actually that useful yet because it's difficult to access the JS console while the browser tab is busy.

I'll simplify this for now.